pkg/crc/api/events/status_change_stream.go

```diff
@@ -0,0 +1,69 @@
+package events
+
+import (
+	"encoding/json"
+
+	"github.com/crc-org/crc/v2/pkg/crc/logging"
+	"github.com/crc-org/crc/v2/pkg/crc/machine"
+	"github.com/crc-org/crc/v2/pkg/crc/machine/state"
+	"github.com/crc-org/crc/v2/pkg/crc/machine/types"
+	"github.com/crc-org/crc/v2/pkg/events"
+	"github.com/r3labs/sse/v2"
+)
+
+type statusChangeEvent struct {
+	Status *types.ClusterStatusResult `json:"status"`
+	Error  string                     `json:"error,omitempty"`
+}
+
+type statusChangeListener struct {
+	machineClient machine.Client
+	publisher     EventPublisher
+}
+
+func newStatusChangeStream(server *EventServer) EventStream {
+	return newStream(newStatusChangeListener(server.machine), newEventPublisher(StatusChange, server.sseServer))
+}
+
+func newStatusChangeListener(client machine.Client) EventProducer {
+	return &statusChangeListener{
+		machineClient: client,
+	}
+}
+
+func (st *statusChangeListener) Notify(changedEvent events.StatusChangedEvent) {
+	logging.Debugf("State Changed Event %s", changedEvent)
+	var event statusChangeEvent
+	status, err := st.machineClient.Status()
+	// if we cannot receive actual state, send error state with error description
+	if err != nil {
+		event = statusChangeEvent{Status: &types.ClusterStatusResult{
+			CrcStatus: state.Error,
+		}, Error: err.Error()}
+	} else {
+		// event could be fired, before actual code, which change state is called
+		// so status could contain 'old' state, replace it with state received in event
+		status.CrcStatus = changedEvent.State // override with actual reported state
```

Comment:
Spending more time on it, I understand more the problems this is trying to solve.
One example is when the VM is stopped. `crc status` will succeed and report that crc is stopped, however `crc stop` will return an error and say the machine is already stopped.

Without `status.CrcStatus = changedEvent.State`, when invoking `api/stop`, we would not be reporting an error in the stream, but instead we'd only say the machine is stopped. This would not match what was reported to the `api/stop` call. There are most likely other state transitions where this is needed.

The comment however seems to indicate sometimes `Status` could report a stale status, I haven't seen that in the little testing I did, but I'd be interested in understanding more when this happens? If we don't record this knowledge now when it's more or less fresh in our minds, it will be very difficult to find this back in 6 months from now if we need it :-/

Reply:
ah, there's more details in an older discussion:
> User call start, we fire state change event, before actuality call Start, and if listener is fast enough, it could get old state (basically we could lost starting state in transition from stoped to running states)

Reply:
In my opinion, this is an indication that these events are not fired in the right place, and this is papered over with even more complexity :-/

Reply:
The part you quoted is a description of the behaviour in this PR. These are some notes I took during review/testing when I was trying to understand the design choices which led to the code being reviewed.

Reply:
I try to replace this code, with firing event at more "proper" place, but face problem that `starting` state is "artificial", ie exist only in `sync.go` file, and VM doesn't have/reports `starting` state. Also `stopping` state is the same. 
So I have no other ideas, than accept current PR and think on better state handling, like @cfergeau propose  there https://github.com/crc-org/crc/pull/3903#discussion_r1476135076

Or change state handling first, and then rebase this PR on top of it.

@gbraad @cfergeau WDYT?